proposals/caffe2-operator-proposal.md

```diff
@@ -0,0 +1,134 @@
+## Motivation
+Caffe2 is a popular machine learning framework which currently does not have an operator/controller for Kubernetes. This proposal is aimed at defining what that operator should look like, and adding it to Kubeflow.
+
+For distributed training, Caffe2 has no parameter server compared with Tensorflow, so it has to use Redis/MPI to find the other nodes to communicate. 
+
+## Goals
+A Kubeflow user should be able to run training using Caffe2 as easily as they can using Tensorflow.  This proposal is centered around a Kubernetes operator for Caffe2. A user should be able to run both single node and distributed training jobs with Caffe2.
+
+This proposal defines the following:
+- A Caffe2 operator
+- A way to deploy the operator with kubectl
+- A single pod Caffe2 example
+- A distributed Caffe2 example
+- A distributed Caffe2 proposal with [batchd scheduler](https://github.com/kubernetes-incubator/kube-arbitrator)
+
+## Non-Goals
+For the scope of this proposal, we won't be addressing the method for serving the model.
+
+## API (CRD and resulting objects)
+
+### Custom Resource Definition
+The custom resource submitted to the Kubernetes API would look something like this:
+
+```yaml
+apiVersion: "kubeflow.org/v1alpha1"
+kind: "Caffe2Job"
+metadata:
+  name: "example-job"
+spec:
+  backend: "redis"
+  replicaSpecs:
+    - replicas: 1
+      ReplicaType: MASTER
+      template:
+        spec:
+          hostNetwork: true
+          containers:
+            - image: carmark/caffe2:latest
+              name: caffe2
+              securityContext:
+                capabilities:
+                  add: ["ALL"]
+          restartPolicy: Never
+    - replicas: 2
+      ReplicaType: WORKER
+      template:
+        spec:
+          hostNetwork: true
```

Comment:
What is the reasoning for using hostNetwork?

Reply:
In our case, we use IB to communicate, so the `hostNetwork` is necessary.